#!/bin/bash
#PBS -q hotel
#PBS -N check_beagle_output_variants_%CHROM%
#PBS -l nodes=1:ppn=2
#PBS -l walltime=48:00:00
#PBS -o /projects/ps-gymreklab/jmargoli/ukbiobank/str_imputed/runs/%RUN_NAME%/batches/output
#PBS -e /projects/ps-gymreklab/jmargoli/ukbiobank/str_imputed/runs/%RUN_NAME%/batches/output
#PBS -V
#PBS -M jmargoli@ucsd.edu
#PBS -m a

#This should take a bit over half a day
#For comparing a UKB callset with all individuals
#against the reference panel

run_name=%RUN_NAME%
chrom=%CHROM%

#enable running via command line
if [ ! -z "$1" ] ; then
	if [ -z "$2" ] ; then
		echo "Error: expected two arguments - chrom and run_name" 1>&2
		exit -1
	fi
	run_name=$1
	chrom=$2
fi

echo CHROM $chrom RUN_NAME $run_name
>&2 echo CHROM $chrom RUN_NAME $run_name

source ~/.bashrc
conda activate ukb_analysis
files=$(python $UKB/str_imputed/scripts/list_batches_in_order.py $run_name $chrom)
conda deactivate

if [ -z "$files" ] ; then
	echo "Error: Found no files matching $UKB/str_imputed/runs/$run_name/batches/chr${chrom}_samples_*.vcf.gz" 1>&2
	exit -1
fi

for file in $files ; do
	echo "Checking file $file"
	time cmp <( cut -f1-5 $UKB/snpstr/vcf_1_sample/chr$chrom.vcf | grep -v '^#' ) \
	    <( zcat $file | cut -f1-5 | grep -v '^#' ) || \
	{
		echo "The variants in file $file do not match the reference panel" 1>&1
		exit -1
	}
done

echo "Success. All variants for chromosome $chrom match those in the reference panel."

